File: frontend/admin/src/services/roleService.ts
import httpClient from "@/configurations/httpClient";
import type { Permission } from "./permissionService";
import {handleApiResponse } from "@/utils/apiResponse";
import type { ApiResponse } from "@/utils/apiResponse";

export interface Role {
  name: string;
  description: string;
  permissions: Permission[];
}

export interface RoleRequest {
  name: string;
  description: string;
  permissions: string[]; // Array of permission names
}


// Get all roles
export const getAllRoles = async (): Promise<Role[]> => {
  return handleApiResponse<Role[]>(
    httpClient.get<ApiResponse<Role[]>>("/roles")
  );
};

// Create role
export const createRole = async (request: RoleRequest): Promise<Role> => {
  return handleApiResponse<Role>(
    httpClient.post<ApiResponse<Role>>("/roles", request)
  );
};

// Update role
export const updateRole = async (
  roleId: string,
  request: RoleRequest
): Promise<Role> => {
  const response = await httpClient.put<ApiResponse<Role>>(
    `/roles/${roleId}`,
    request
  );
  return response.data.result;
};

// Delete role
export const deleteRole = async (roleId: string): Promise<void> => {
  await httpClient.delete(`/roles/${roleId}`);
};

File: backend/theatermgnt/src/main/java/com/theatermgnt/theatermgnt/authorization/service/RoleService.java
package com.theatermgnt.theatermgnt.authorization.service;

import java.util.HashSet;
import java.util.List;

import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;

import com.theatermgnt.theatermgnt.authorization.dto.request.RoleRequest;
import com.theatermgnt.theatermgnt.authorization.dto.response.RoleResponse;
import com.theatermgnt.theatermgnt.authorization.entity.Role;
import com.theatermgnt.theatermgnt.authorization.mapper.RoleMapper;
import com.theatermgnt.theatermgnt.authorization.repository.PermissionRepository;
import com.theatermgnt.theatermgnt.authorization.repository.RoleRepository;
import com.theatermgnt.theatermgnt.common.exception.AppException;
import com.theatermgnt.theatermgnt.common.exception.ErrorCode;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
@Slf4j
@PreAuthorize("hasRole('ADMIN')")
public class RoleService {
    RoleRepository roleRepository;
    PermissionRepository permissionRepository;
    RoleMapper roleMapper;

    /// CREATE A ROLE
    public RoleResponse create(RoleRequest request) {
        var role = roleMapper.toRole(request);

        var permissions = permissionRepository.findAllById(request.getPermissions());
        role.setPermissions(new HashSet<>(permissions));

        role = roleRepository.save(role);
        return roleMapper.toRoleResponse(role);
    }

    /// GET ALL ROLES
    public List<RoleResponse> getAll() {
        var roles = roleRepository.findAll();
        return roles.stream().map(roleMapper::toRoleResponse).toList();
    }

    /// UPDATE A ROLE
    public RoleResponse update(String roleId, RoleRequest request) {
        Role role = roleRepository.findById(roleId).orElseThrow(() -> new AppException(ErrorCode.ROLE_NOT_FOUND));
        roleMapper.updateRole(request, role);

        var permissions = permissionRepository.findAllById(request.getPermissions());
        role.setPermissions(new HashSet<>(permissions));
        role = roleRepository.save(role);
        return roleMapper.toRoleResponse(role);
    }

    /// DELETE A ROLE
    public void delete(String roleId) {
        roleRepository.deleteById(roleId);
    }
}
